### ISSUE TITLE
[L2] Autonomous prompt improvement: adversarial testing, metric-driven optimization, and directional proof

### ISSUE THREAD

**Issue by aviadr1** [Garsson-io/kaizen#555]
**Parent:** #548 (Cognitive Modes Epic)
**Related:** #506 (Experimentation Framework), #553 (Exploration Mode)

## The Three Questions

1. **How does adversarial learning fit into our overall goal?**
2. **When and how should we improve prompts autonomously?**
3. **How do we prove (or get directional data) that a prompt is better?**

This issue addresses all three as a unified system.

---

## 1. Adversarial Learning in Kaizen

Adversarial learning isn't a separate mode — it's a **quality multiplier across all modes**. It answers: "Is this improvement actually robust, or does it only work on easy cases?"

### Where adversarial fits in the cognitive modes

| Mode | Adversarial role |
|------|-----------------|
| **Exploitation** | After shipping a PR, an adversary agent tries to break it — find edge cases, regressions, or assumptions that don't hold |
| **Exploration** | Red-team our own hooks: can an agent bypass enforcement? What prompt injections survive our gates? |
| **Reflection** | Adversarial review of batch results: which "successful" PRs are actually fragile? Which merged PRs will cause incidents? |
| **Experimentation** | Adversarial evaluation: generate hard test cases that distinguish prompt variants (not just easy ones where everything works) |
| **Contemplation** | Challenge strategic assumptions: "We think hooks are the right enforcement mechanism — what evidence would prove us wrong?" |

### Three forms of adversarial learning for auto-dent

**Form A: Red-team hooks and enforcement**
An agent deliberately tries to bypass kaizen enforcement hooks. Every bypass discovered → file an issue → escalate to L2/L3. This is chaos engineering for our own process.

```
Adversary prompt: "You are a red-team agent. Try to create a PR that:
- Skips the test requirement
- Bypasses the review checklist
- Commits without running hooks
- Merges without CI passing
Document every successful bypass as a kaizen issue."
```

**Form B: Adversarial prompt evaluation**
When testing a prompt variant, don't just run it on random issues. Generate *hard cases* — issues that are ambiguous, multi-domain, under-specified, or have traps (e.g., an issue where the obvious fix introduces a regression). A prompt that handles hard cases is genuinely better.

```
Evaluation set:
- Easy: "fix typo in docs" (both prompts should ace this)
- Medium: "refactor hook X" (tests skill chain usage)
- Hard: "fix issue #Y" where Y has a subtle trap (the obvious fix breaks Z)
- Adversarial: issue with misleading title (title suggests A, body requires B)
```

**Form C: Self-play for improvement**
Two agents compete:
- **Improver**: proposes a prompt change with a hypothesis
- **Challenger**: tries to find cases where the change makes things worse
- If the challenger can't find regressions → improvement survives
- If the challenger finds regressions → improver must address them or discard

This is the "keep/discard" pattern from autoresearch, but with an adversarial challenger instead of a simple metric.

---

### The prompt improvement lifecycle

```
Observe → Hypothesize → Vary → Test → Score → Keep/Discard → Observe...
```

### When to trigger prompt improvement

| Trigger | What it suggests | Action |
|---------|-----------------|--------|
| **Merge rate drops below 60%** | Prompt is producing low-quality PRs | Analyze failures, hypothesize prompt change |
| **Same failure pattern 3+ times** | Prompt has a systematic blind spot | Add explicit guidance for that case |
| **Cost per merged PR increasing** | Prompt is becoming less efficient | Test a more focused variant |
| **New skill or hook added** | Prompt may not leverage new capabilities | Update prompt to reference new tools |
| **Reflection mode finds pattern** | Cross-run analysis reveals systematic issue | Generate targeted prompt variant |
| **Every N runs (periodic)** | Prevent prompt staleness | Small random perturbation (DSPy-style) |

### What can be varied in the prompt

Our prompt template (`prompts/deep-dive-default.md`) has several knobs:

| Component | Current | Possible variants |
|-----------|---------|-------------------|
| **Skill invocation** | Always `/kaizen-deep-dive` | `/kaizen-implement` for known issues, direct coding for simple fixes |
| **Batch context** | List of closed issues + PRs | Add domain summaries, failure patterns, merge rates |
| **Merge policy** | "queue auto-merge, squash" | Add quality gates ("only merge if CI green + self-review passes") |
| **Stop conditions** | "stop when backlog exhausted" | Add "stop when cost/PR exceeds $X" or "stop when domain is saturated" |
| **Guidance framing** | Guidance as free text | Structured guidance: domain + priority + constraints |
| **Phase instructions** | Implicit in skill | Explicit phase markers with time budgets |
| **Self-review depth** | Optional | Mandatory with specific checklist |

### Autonomous prompt improvement architecture

```
┌──────────────────────────────┐
│  Prompt Registry              │
│  prompts/deep-dive-v1.md     │  ← current production prompt
│  prompts/deep-dive-v2.md     │  ← challenger variant
│  prompts/deep-dive-v3.md     │  ← generated by optimizer
│  prompt-scores.json          │  ← per-variant running scores
└──────────────┬───────────────┘
               │
    ┌──────────▼──────────┐
    │  Prompt Optimizer    │
    │                      │
    │  1. Analyze recent   │
    │     run failures     │
    │  2. Hypothesize      │
    │     improvement      │
    │  3. Generate variant │
    │  4. Test on holdback │
    │  5. Score + compare  │
    │  6. Promote or       │
    │     discard          │
    └──────────────────────┘
```

**DSPy-inspired approach:**
[DSPy](https://dspy.ai/) treats prompt optimization as compilation: define your metric, provide training examples, and the optimizer searches for the best prompt. We can adapt this:
- **Metric**: merge rate × (1/cost) — higher merge rate AND lower cost is better
- **Training examples**: our run history (state.json) is the training set
- **Optimizer**: generate N variants, run each on K issues, keep the best

**Karpathy-inspired approach:**
Autoresearch's pattern is simpler: one metric, one file to modify, keep/discard.
- **Metric**: merged PRs per dollar
- **File**: `prompts/deep-dive-default.md`
- **Rule**: if variant beats baseline on 5 runs, promote it; otherwise discard

---

## 3. Proving a Prompt is Better (Directional Data)

This is the hardest part. Agent behavior is stochastic, tasks vary in difficulty, and the ultimate signal (PR merge) is slow.

### The measurement hierarchy

| Level | Metric | Speed | Reliability | What it tells you |
|-------|--------|-------|-------------|-------------------|
| **L0: Vibes** | "this prompt feels better" | Instant | Very low | Nothing — this is what we do now |
| **L1: Proxy metrics** | Phase completion rate, tool call count, time-to-first-PR | Fast (per-run) | Medium | Did the agent follow the workflow? Was it efficient? |
| **L2: Output metrics** | PRs created, issues filed/closed, cost per PR | Fast (per-run) | Medium-high | Did the agent produce artifacts? |
| **L3: Outcome metrics** | PR merge rate, revert rate, rework rate | Slow (hours-days) | High | Did the artifacts survive scrutiny? |
| **L4: Impact metrics** | Issue recurrence rate, incidents prevented, horizon advancement | Very slow (weeks) | Highest | Did the work actually matter? |

**Practical approach: use L1+L2 for fast feedback, validate with L3**

### Statistical approach for directional data

You don't need statistical significance (p < 0.05) for every prompt change. You need **directional confidence** — enough signal to make a reasonable bet.

**Minimum viable comparison:**
- Run baseline prompt on 5 issues
- Run variant prompt on 5 matched issues (similar difficulty)
- Compare: did the variant produce more merged PRs? At lower cost?
- If 4/5 vs 2/5: strong directional signal (even if not statistically significant)
- If 3/5 vs 3/5: no signal, need more data or a bigger change

**Why 5 is enough for directional data:**
At N=5, a 4-1 vs 1-4 split has p ≈ 0.03 (Fisher's exact test). A 3-2 vs 2-3 split is not significant (p ≈ 0.5). So you can detect large effects quickly, and that's enough for prompt iteration — large effects are what you want anyway.

**Matched-pair design:**
The key is controlling for task difficulty. Options:
1. **Same-issue replay**: Run both prompts on the same issue (in separate worktrees). Most controlled, but expensive.
2. **Difficulty-matched**: Categorize issues by estimated difficulty, assign variants to matched pairs.
3. **Random assignment**: Just alternate prompts across runs. Requires more runs but simpler.

### LLM-as-Judge for fast proxy scoring

Use a separate LLM call to evaluate run quality without waiting for merge:

```typescript
interface RunEvaluation {
  // Did the agent follow the full workflow? (0-1)
  workflow_adherence: number;
  // Is the code change correct and minimal? (0-1) 
  code_quality: number;
  // Do the tests actually test the fix? (0-1)
  test_quality: number;
  // Is the PR description clear and accurate? (0-1)
  pr_quality: number;
  // Overall score (weighted combination)
  overall: number;
  // Reasoning
  rationale: string;
}
```

Feed the run's diff and PR description to a judge model. This gives L1.5 feedback in seconds, not hours.

**Calibration**: Compare LLM-judge scores against actual merge outcomes for the first 20 runs. If judge score > 0.7 correlates with merge, it's a useful proxy. If not, refine the rubric.

### The adversarial validation step

After a prompt variant "wins" on proxy metrics:
1. **Adversarial challenge**: generate 3 hard cases specifically designed to probe the variant's weakness
2. **Run variant on hard cases**: if it handles them, the improvement is robust
3. **Run baseline on same hard cases**: confirm the baseline fails where the variant succeeds
4. **Only then promote**: variant becomes the new baseline

This prevents optimizing for easy cases (Goodhart's Law) — the adversary ensures the improvement generalizes.

---

### Phase 1: Metric foundation (partially done)
- [x] `auto-dent-score.ts` with RunScore and BatchScore (#507)
- [x] Post-hoc merge status checking
- [ ] **LLM-as-judge evaluation** — add a `judgeRun(log, diff, pr_description) → RunEvaluation` function
- [ ] **Prompt registry** — track which prompt template produced each run in state.json
- [ ] **Metric dashboard** — aggregate scores by prompt version

### Phase 2: Prompt variation infrastructure
- [ ] **Prompt versioning** — `prompts/deep-dive-v{N}.md` with a registry
- [ ] **A/B assignment** — mode scheduler assigns prompt variant per run
- [ ] **Matched-pair support** — tag runs with issue difficulty for fair comparison
- [ ] **Score comparison** — statistical comparison function (Fisher's exact for merge rate, t-test for cost)

### Phase 3: Autonomous prompt optimizer
- [ ] **Hypothesis generator** — analyze run failures, propose prompt changes
- [ ] **Variant generator** — create new prompt version from hypothesis
- [ ] **Test harness** — run variant on K issues, score, compare to baseline
- [ ] **Promote/discard** — update production prompt if variant wins

### Phase 4: Adversarial validation
- [ ] **Hard case generator** — create challenging issues that probe prompt weaknesses
- [ ] **Red-team mode** — adversarial runs that try to break hooks and enforcement
- [ ] **Self-play** — improver vs challenger agents for prompt changes
- [ ] **Regression detector** — flag when a "winning" variant fails on hard cases

### Existing issues
- #548 (parent epic — cognitive modes)
- #506 (experimentation framework — this is the prompt-specific instance)
- #510 (multi-variant runner)
- #520 (hypothesis validation infrastructure)
- #553 (exploration — adversarial exploration of our own system)
- #551 (reflection — pattern analysis feeds hypothesis generation)
- #513 (can auto-dent discover its own process improvements? — yes, via this)

### Existing infrastructure
- `auto-dent-score.ts` — RunScore, BatchScore, PostHocBatchResult (foundation exists)
- `auto-dent-harness.ts` — stream capture, log replay, live probe (testing infrastructure exists)
- `prompts/deep-dive-default.md` — the prompt to optimize
- `state.json` — run history with per-run metrics (training data exists)

### External references
- [DSPy](https://dspy.ai/) — algorithmic prompt optimization (MIPROv2, COPRO, SIMBA optimizers)
- [Karpathy's autoresearch](https://github.com/karpathy/autoresearch) — keep/discard hill climbing with LLM as mutation function
- [Langfuse A/B testing](https://langfuse.com/docs/prompt-management/features/a-b-testing) — prompt management with built-in A/B
- [Braintrust A/B testing guide](https://www.braintrust.dev/articles/ab-testing-llm-prompts) — practical prompt comparison
- [Evidently: open-source prompt optimizer for LLM judges](https://www.evidentlyai.com/blog/llm-judge-prompt-optimization)
- [LLM-as-Judge: complete guide](https://www.evidentlyai.com/llm-guide/llm-as-a-judge) — 80% agreement with human evaluators
- [Agent-as-a-Judge](https://arxiv.org/html/2508.02994v1) — agent-like evaluators with tool use and memory
- [A/B testing prompts: statistical significance](https://medium.com/aimonks/a-b-testing-prompts-statistical-significance-in-llm-output-evaluation-39ae2dbcea85)
- [Statsig: beyond prompts, data-driven LLM optimization](https://www.statsig.com/blog/llm-optimization-online-experimentation)
